syntax = "proto3";
import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";

package P2PFileSync;

enum ProtoPayloadType{
  HELLO = 0;
  REDIRECT = 7;
  ACK = 5;
  NACK = 6;
  PING = 1;
  PONG = 2;
  HEARTBEAT = 4;
  FILE_REQUEST = 3;
}

enum ProtoSignAlgorithm{
  SHA_256 = 0;
}

message ProtoMessage {
  // sender/receiver information
  bytes sender = 5;
  bytes receiver = 6;

  // packet information
  ProtoPayloadType packet_type = 1;
  google.protobuf.Timestamp timestamp = 3;

  // actual payload
  google.protobuf.Any payload = 2;
}

message SignedProtoMessage{
  // sender/receiver information
  bytes sender = 1;
  bytes receiver = 2;
  uint32 last_forward_port = 3;

  // packet information
  uint32 ttl = 4;

  // signature part
  bytes signature = 5;
  bytes signed_payload = 6;
  ProtoSignAlgorithm sign_algorithm = 7;

}

message ProtoHelloMessage {
  bytes sender_id = 5;
  bytes verified_msg = 6;
  bytes x509_certificate = 2;
}